### ISSUE TITLE
[New build - EXPERIMENTAL] 22w30a: Furnace, Conveyor, Collector, Inventory

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#4778]
This build introduces several features to prepare the addition of crafting to the game:
functional cargo, conveyor tube with auto pull and push logic, collector, conveyor belt and finally furnace and ingot smelting.

Functional Cargo:
It is now possible to store items inside cargo container.
It is possible to rename container from their UI.

Conveyor Tubes:
Straight, angled and 6-way conveyor tubes where added. 
Connects inventory block with them to do automatic or manual transfer.
An indicator light have been added to help you visualize what is happening:
 - Red: no connection
 - Green: valid connection
 - Flashing: item being transferred
There is a maximum limit of 6 connections for cargo and other inventory blocks.
The grid scale impact the rate at which items are being transferred, the larger the faster.

Push and Pull Logic:
Some inventory blocks are able to push or pull blocks.
The collector will push blocks when set at collect, and pull block when set at expel.
Custom rules will be added to give a finer control of what item can be pushed / pulled.

Collector:
The collector can pull and collect nearby items or can be set to expel its inside inventory.
The range at which it will pull object is also configurable.

Furnace and Smelting:
A furnace block has been added.
It will pull ore placed close to its entry.
It can bet set to expel crafted ingots to its output, so you can link it will conveyor belts.

Conveyor Belt:
The conveyor belt will carry items placed on top of it.
It can push player or brick entity too.
It uses real physics, so item can fall!
It can be scaled horizontally.
A sloped and angled belt will be added.

This build breaks the save compatibility for item stands and player item bar.

Hotfixes:
Light sprite on handheld does not appear at the right place.

**Comment by tsunamayo** [comment]
Build is up! It is quite a big one, so I am sure I will have plenty of things to fix. Thanks for your feedback!

**Comment by daOyster** [comment]
This is awesome. Random question, if conveyors can push players, would it be possible to apply similar logic to allow players to stick to moving Child entities instead of having no player friction with them? That could allow us to build horizontal trams using pistons/sliders that the player won't slide around in/off once it starts moving.

**Comment by PopkaAkula** [comment]
it says 22w301 in-game btw

![image](https://user-images.githubusercontent.com/37738021/181787497-ac287f4d-6872-4ca5-ab41-37d7735042b1.png)

**Comment by daOyster** [comment]
Also one other idea if it isn't already a thing, can we stop/start the conveyors with a on/off logic signal?

**Comment by Asnro** [comment]
People on discord raised the concern of conveyor belts causing lag or items glitching out of existence due to the transported objects still having physics. Would it be possible to copy Satisfactory's belts, where objects are "glued" to the belt and won't bounce around? This might reduce the physics lag in large factories and on weak computers, and will ensure objects won't fall off.

**Comment by pinesh** [comment]
Exciting stuff. Absolutely not what I was expecting!

**Comment by Dwarf-LordPangolin** [comment]
This is probably one of the parts of the game that I'm the most excited about: the crafting aspect!

Before development goes any further though, I would **_strongly_** recommend against conveyors with actual objects on them. That's going to cause lots of problems in multiple ways.

1. Everyone has already mentioned the performance issues caused by physics, but nobody has talked about conveyors that are set up in zero gravity: stuff can just float away there.
2. As the demonstration video that was posted on Steam shows, things can fall off. Fallout 4 has the same problem, and it is nothing but frustration: you can't rely on your manufacturing systems because you keep having to babysit them and clean things off the floor, and it's a huge quality of life problem.
3. Simulating the movement of each individual object as it moves from one inventory to another, even if there are no physics involved, is a performance overhead with no benefit. For performance reasons, if an object is in one inventory, and another connected inventory needs it, it should just be moved instantly there. Simulating the actual movement achieves nothing and will end up costing a lot, because the factories that players make are going to be **_huge_**. If you thought players went nuts with child entities, just wait till you see what they do with manufacturing! 😱 
4. On a final related note, just as there is no gameplay reason to have discrete objects, regardless of whether they have physics simulations or not, visible on the belts, there's also no reason the conveyor should even have a visible belt either. As long as the connected blocks' inventories can talk to each other, that's all that's needed. None of it needs to be visible. My suggestion would be to adopt an enclosed conveyor system, similar to those used in real life like these:
![_klein_107_002-1600_2](https://user-images.githubusercontent.com/33041786/181827557-f254f3c3-404a-49c7-a6c8-75b344c39c7f.jpg)
Or the ones used in Space Engineers:
![space-engineers-heavy-industry-dlc-is-available-now](https://user-images.githubusercontent.com/33041786/181829365-b0b5a004-977a-42ee-ac3c-f04f93be8981.jpg)

None of the items are visible, and objects are instantly transferred. The same results, with vastly less performance overhead. The only real benefit that visible objects moving on visible belts achieves is that it looks nice, which isn't a bad reason, but I don't feel it's enough to justify the issues it brings with it.

I haven't had a chance to test this yet, and I am massively excited about this! But I can also already tell you, based on experience with other games, that having little objects traveling around on moving belts is going to give you **_huge_** headaches in the future.

**Comment by MagicalAbyss** [comment]
While real physics are nice and all, but have you stop a moment and think what happen to the game performances with mega factories? Imagine hundreds if not thousands of ores/ingots just moving through conveyor belts. Please have options to disable physics or even better with "enclosed conveyors" that do not show objects at all.

**Comment by CurioInventorium** [comment]
Awesome to see things moving along (no pun intended). Just need to raise the issue of colour blindness since I'm red/green colour blind. I'd like to suggest the red indicator for the conveyor tube have some sort of pattern with the red, like angled black lines. Or instead of red, just have a black indicator. It would really help, thank you! 😄

**Comment by Dwarf-LordPangolin** [comment]
> While real physics are nice and all, but have you stop a moment and think what happen to the game performances with mega factories? Imagine hundreds if not thousands of ores/ingots just moving through conveyor belts. Please have options to disable physics or even better with "enclosed conveyors" that do not show objects at all.

I had the same concerns, and actually wrote up a post detailing them, but then I deleted it after someone reminded me that there are also enclosed conveyor tubes. Assuming those provide the same functionality as the belt versions, those should take care of any possible issues.

**Comment by cloudrambler** [comment]
> Awesome to see things moving along (no pun intended). Just need to raise the issue of colour blindness since I'm red/green colour blind. I'd like to suggest the red indicator for the conveyor tube have some sort of pattern with the red, like angled black lines. Or instead of red, just have a black indicator. It would really help, thank you! 😄

Aesthetically, having red indicators when there is nothing "wrong" is a little counterintuitive. So I second the idea of the disconnected state being black. I would go one step further and allow the connected state to be adjustable; perhaps with the RGB logic brick.

**Comment by tsunamayo** [comment]
> This is awesome. Random question, if conveyors can push players, would it be possible to apply similar logic to allow players to stick to moving Child entities instead of having no player friction with them? That could allow us to build horizontal trams using pistons/sliders that the player won't slide around in/off once it starts moving.

Yes I will try to use the same logic. Player control and collision needs a big rework.

**Comment by tsunamayo** [comment]
> > While real physics are nice and all, but have you stop a moment and think what happen to the game performances with mega factories? Imagine hundreds if not thousands of ores/ingots just moving through conveyor belts. Please have options to disable physics or even better with "enclosed conveyors" that do not show objects at all.
> 
> I had the same concerns, and actually wrote up a post detailing them, but then I deleted it after someone reminded me that there are also enclosed conveyor tubes. Assuming those provide the same functionality as the belt versions, those should take care of any possible issues.

Yes exactly, I wanted belts and physics because those are fun and fun is what I am lacking, but you can still use closed tubes if you prefer. Also this is not satisfactory, so you should not have that much item to process.
But more optimizations are on the way.

**Comment by tsunamayo** [comment]
> > Awesome to see things moving along (no pun intended). Just need to raise the issue of colour blindness since I'm red/green colour blind. I'd like to suggest the red indicator for the conveyor tube have some sort of pattern with the red, like angled black lines. Or instead of red, just have a black indicator. It would really help, thank you! 😄
> 
> Aesthetically, having red indicators when there is nothing "wrong" is a little counterintuitive. So I second the idea of the disconnected state being black. I would go one step further and allow the connected state to be adjustable; perhaps with the RGB logic brick. Of course it can be green by default.

Yes indeed it makes sense!
